rosrust/src/tcpros/service.rs

```diff
@@ -157,7 +172,7 @@ where
     // TODO: validate message length
     let _length = stream.read_u32::<LittleEndian>();
     // Break out of loop in case of failure to read request
-    while let Ok(req) = RosMsg::decode(&mut stream) {
+    if let Ok(req) = RosMsg::decode(&mut stream) {
```

Comment:
Aren't clients able to send multiple requests via the same connection to a service?
That method greatly reduces communication overhead, since header exchanges can often be much larger than the actual service payloads.

Reply:
Yes, good catch, you're right this does prevent persistent connections.

Looks like [service_client_link.cpp](https://github.com/ros/ros_comm/blob/89ca9a7bf288b09d033a1ab6966ef8cfe39e1002/clients/roscpp/src/libros/service_client_link.cpp) is where roscpp handles these things. I see two differences from service.rs:
1. It looks for "persistent" in the header to know whether there might be more messages.
2. It reads the length of the message before each new message is deserialized.

Reply:
Could you just leave a TODO above the if line, and I can take on handling both cases

Reply:
Nvm, I've just made an issue and will handle it afterwards.